### PR TITLE
added support for overriding manifest annotations

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Hello, this PR will add overriding manifest annotations support to crane's `mutate` command, it will work like the following picture:

Screen Shot 2021-06-18 at 01 06 08

### CODE REVIEW

File: cmd/crane/cmd/mutate.go [hunk 1]
```diff
@@ -100,8 +110,21 @@ func NewCmdMutate(options *[]crane.Option) *cobra.Command {
...
 		},
 	}
+	mutateCmd.Flags().StringSliceVarP(&anntns, "annotation", "a", nil, "New annotations to add")
 	mutateCmd.Flags().StringSliceVarP(&lbls, "label", "l", nil, "New labels to add")
 	mutateCmd.Flags().StringVar(&entrypoint, "entrypoint", "", "New entrypoing to set")
 	mutateCmd.Flags().StringVarP(&newRef, "tag", "t", "", "New tag to apply to mutated image. If not provided, push by digest to the original image repository.")
 	return mutateCmd
 }
+
+func splitKeyValuePairsInFormOfAnEqualSign(kvPairs []string) (map[string]string, error) {
```
Comment (imjasonh) [inline cmd/crane/cmd/mutate.go:120]: This is a bit wordy. How about just `splitKeyVals` and a doc comment?

File: pkg/v1/mutate/mutate.go [hunk 2]
```diff
@@ -113,6 +113,29 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
 	return ConfigFile(base, cf)
 }
 
+// ManifestAnnotations mutates the provided v1.Image to have the provided annotations
+func ManifestAnnotations(base v1.Image, annotations map[string]string) (v1.Image, error) {
+	m, err := base.Manifest()
+	if err != nil {
+		return nil, err
+	}
```
Comment (imjasonh) [inline pkg/v1/mutate/mutate.go:121]: Can we `m.DeepCopy` here instead, as early as possible and before any changes?

File: pkg/v1/mutate/mutate_test.go [hunk 3]
```diff
@@ -266,6 +266,31 @@ func TestMutateConfig(t *testing.T) {
...
+func TestMutateManifestAnnotations(t *testing.T) {
+	source := sourceImage(t)
+
+	newAnnotations := map[string]string{
+		"im.the.first.annotation": "hello world",
+	}
+
+	result, err := mutate.ManifestAnnotations(source, newAnnotations)
+	if err != nil {
+		t.Fatalf("failed to mutate a config: %v", err)
```
Comment (imjasonh) [inline pkg/v1/mutate/mutate_test.go:278]: "failed to mutate annotations"

File: pkg/v1/mutate/mutate_test.go [hunk 4]
```diff
@@ -266,6 +266,31 @@ func TestMutateConfig(t *testing.T) {
...
+	}
+
+	if manifestsAreEqual(t, source, result) {
```
Comment (imjasonh) [inline pkg/v1/mutate/mutate_test.go:281]: Does this indicate that the source was mutated in the process? I would expect source to be unmodified and the modified result to be returned.

File: pkg/v1/mutate/mutate_test.go [hunk 5]
```diff
@@ -266,6 +266,31 @@ func TestMutateConfig(t *testing.T) {
...
+		t.Error("mutating the manifest annotations MUST mutate the manifest")
+	}
+
+	if configDigestsAreEqual(t, source, result) {
+		t.Errorf("mutating the manifest annotations MUST mutate the config digest")
```
Comment (imjasonh) [inline pkg/v1/mutate/mutate_test.go:286]: This actually should prove that the config isn't modified while modifying annotations.

**Review by imjasonh — COMMENTED** [review]
Thanks for this change! 🎉

**Comment by jonjohnsonjr** [comment]
I'll add some context from https://github.com/google/go-containerregistry/issues/281 we talked about how this would work for both image and manifest lists. I don't think you need to solve this immediately, but being able to annotate just the top-level manifest vs recursing and annotating every manifest is an interesting problem.

File: pkg/v1/mutate/mutate.go [hunk 6]
```diff
@@ -113,6 +113,29 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
 	return ConfigFile(base, cf)
 }
 
+// ManifestAnnotations mutates the provided v1.Image to have the provided annotations
+func ManifestAnnotations(base v1.Image, annotations map[string]string) (v1.Image, error) {
```
Comment (jonjohnsonjr) [inline pkg/v1/mutate/mutate.go:117]: I'd just call this `Annotations` because most of the other functions assume an image as well. If we want to extend this to other things, we can follow the `IndexMediaType` pattern and just do `IndexAnnotations`.

Ideally, this function would take a Describable, but then it's hard to know what to return :(

File: pkg/v1/mutate/mutate_test.go [hunk 7]
```diff
@@ -266,6 +266,31 @@ func TestMutateConfig(t *testing.T) {
 	}
 }
 
+func TestMutateManifestAnnotations(t *testing.T) {
```
Comment (jonjohnsonjr) [inline pkg/v1/mutate/mutate_test.go:269]: Just call this `TestAnnotations`.

File: pkg/v1/mutate/mutate.go [hunk 8]
```diff
@@ -113,6 +113,29 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
...
+	m, err := base.Manifest()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(m.Annotations) < 1 {
```
Comment (jonjohnsonjr) [inline pkg/v1/mutate/mutate.go:123]: Check for `nil` instead?

File: pkg/v1/mutate/mutate.go [hunk 9]
```diff
@@ -113,6 +113,29 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
...
+	for k, v := range annotations {
+		m.Annotations[k] = v
+	}
+
+	image := &image{
+		base:     base,
+		manifest: m.DeepCopy(),
+	}
+
+	return image, err
```
Comment (jonjohnsonjr) [inline pkg/v1/mutate/mutate.go:136]: `return image, nil`

File: pkg/v1/mutate/mutate_test.go (hunk 5 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate_test.go:282]: yes you are right, thank you so much for the feedbacks 🙏

File: pkg/v1/mutate/mutate_test.go (hunk 3 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate_test.go]: thank you 🤦

File: pkg/v1/mutate/mutate.go (hunk 6 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate.go]: done, thank you so much 🙏

File: pkg/v1/mutate/mutate.go (hunk 8 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate.go]: yes, it should be like this way, it is my bad, thank you 🙏

File: pkg/v1/mutate/mutate.go (hunk 9 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate.go]: sorry again my bad 🤦

File: pkg/v1/mutate/mutate_test.go (hunk 7 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate_test.go]: done, thank you 🙏

File: pkg/v1/mutate/mutate.go (hunk 2 above)
Comment (developer-guy) [reply pkg/v1/mutate/mutate.go:121]: thank you so much for this review, but I tried to do that but it didn't work for me, I updated the code with the following content, what did I do wrong?
```go
mn := m.DeepCopy()

if mn.Annotations == nil {
	mn.Annotations = map[string]string{}
}

for k, v := range annotations {
	mn.Annotations[k] = v
}

image := &image{
	base:     base,
	manifest: mn,
}
```

File: cmd/crane/cmd/mutate.go (hunk 1 above)
Comment (developer-guy) [reply cmd/crane/cmd/mutate.go]: done, thank you so much 🙏

File: pkg/v1/mutate/mutate.go [hunk 10]
```diff
@@ -113,6 +113,29 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
 	return ConfigFile(base, cf)
 }
 
+// IndexAnnotations mutates the provided v1.Image to have the provided annotations
+func IndexAnnotations(base v1.Image, annotations map[string]string) (v1.Image, error) {
```
Comment (jonjohnsonjr) [inline pkg/v1/mutate/mutate.go:117]: Sorry, my comment wasn't clear, just call this `Annotations`

Comment (developer-guy) [reply pkg/v1/mutate/mutate.go]: done.

**Comment by jonjohnsonjr** [comment]
Oh yeah, you'll want to do this lazily. That `manifest` field in `image` is the final result of computing everything, so we need to move this logic into `compute()`:

```
$ git diff
diff --git a/pkg/v1/mutate/image.go b/pkg/v1/mutate/image.go
index 14b9c70a..23635f2b 100644
--- a/pkg/v1/mutate/image.go
+++ b/pkg/v1/mutate/image.go
@@ -30,12 +30,13 @@ type image struct {
        base v1.Image
        adds []Addendum
 
-       computed   bool
-       configFile *v1.ConfigFile
-       manifest   *v1.Manifest
-       mediaType  *types.MediaType
-       diffIDMap  map[v1.Hash]v1.Layer
-       digestMap  map[v1.Hash]v1.Layer
+       computed    bool
+       configFile  *v1.ConfigFile
+       manifest    *v1.Manifest
+       annotations map[string]string
+       mediaType   *types.MediaType
+       diffIDMap   map[v1.Hash]v1.Layer
+       digestMap   map[v1.Hash]v1.Layer
 }
 
 var _ v1.Image = (*image)(nil)
@@ -138,6 +139,15 @@ func (i *image) compute() error {
                        manifest.MediaType = *i.mediaType
                }
        }
+       if i.annotations != nil {
+               if manifest.Annotations == nil {
+                       manifest.Annotations = map[string]string{}
+               }
+
+               for k, v := range i.annotations {
+                       manifest.Annotations[k] = v
+               }
+       }
 
        i.configFile = configFile
        i.manifest = manifest
diff --git a/pkg/v1/mutate/mutate.go b/pkg/v1/mutate/mutate.go
index 7615b323..70df4e46 100644
--- a/pkg/v1/mutate/mutate.go
+++ b/pkg/v1/mutate/mutate.go
@@ -115,22 +115,9 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
 
 // Annotations mutates the provided v1.Image to have the provided annotations
 func Annotations(base v1.Image, annotations map[string]string) (v1.Image, error) {
-       m, err := base.Manifest()
-       if err != nil {
-               return nil, err
-       }
-
-       if m.Annotations == nil {
-               m.Annotations = map[string]string{}
-       }
-
-       for k, v := range annotations {
-               m.Annotations[k] = v
-       }
-
        image := &image{
-               base:     base,
-               manifest: m.DeepCopy(),
+               base:        base,
+               annotations: annotations,
        }
 
        return image, nil
```

One thing I'm a little worried about is that if you run this command on a manifest list, I think it will turn that manifest list into an image, because we just call `crane.Pull`. We don't want to do that...

Give me one second, I want to implement something that will let us check if something is an image.

**Comment by jonjohnsonjr** [comment]
> Give me one second, I want to implement something that will let us check if something is an image.

https://github.com/google/go-containerregistry/pull/1057

**Comment by developer-guy** [comment]
> Oh yeah, you'll want to do this lazily. That `manifest` field in `image` is the final result of computing everything, so we need to move this logic into `compute()`:
> 
> ```
> $ git diff
> diff --git a/pkg/v1/mutate/image.go b/pkg/v1/mutate/image.go
> index 14b9c70a..23635f2b 100644
> --- a/pkg/v1/mutate/image.go
> +++ b/pkg/v1/mutate/image.go
> @@ -30,12 +30,13 @@ type image struct {
>         base v1.Image
>         adds []Addendum
>  
> -       computed   bool
> -       configFile *v1.ConfigFile
> -       manifest   *v1.Manifest
> -       mediaType  *types.MediaType
> -       diffIDMap  map[v1.Hash]v1.Layer
> -       digestMap  map[v1.Hash]v1.Layer
> +       computed    bool
> +       configFile  *v1.ConfigFile
> +       manifest    *v1.Manifest
> +       annotations map[string]string
> +       mediaType   *types.MediaType
> +       diffIDMap   map[v1.Hash]v1.Layer
> +       digestMap   map[v1.Hash]v1.Layer
>  }
>  
>  var _ v1.Image = (*image)(nil)
> @@ -138,6 +139,15 @@ func (i *image) compute() error {
>                         manifest.MediaType = *i.mediaType
>                 }
>         }
> +       if i.annotations != nil {
> +               if manifest.Annotations == nil {
> +                       manifest.Annotations = map[string]string{}
> +               }
> +
> +               for k, v := range i.annotations {
> +                       manifest.Annotations[k] = v
> +               }
> +       }
>  
>         i.configFile = configFile
>         i.manifest = manifest
> diff --git a/pkg/v1/mutate/mutate.go b/pkg/v1/mutate/mutate.go
> index 7615b323..70df4e46 100644
> --- a/pkg/v1/mutate/mutate.go
> +++ b/pkg/v1/mutate/mutate.go
> @@ -115,22 +115,9 @@ func Config(base v1.Image, cfg v1.Config) (v1.Image, error) {
>  
>  // Annotations mutates the provided v1.Image to have the provided annotations
>  func Annotations(base v1.Image, annotations map[string]string) (v1.Image, error) {
> -       m, err := base.Manifest()
> -       if err != nil {
> -               return nil, err
> -       }
> -
> -       if m.Annotations == nil {
> -               m.Annotations = map[string]string{}
> -       }
> -
> -       for k, v := range annotations {
> -               m.Annotations[k] = v
> -       }
> -
>         image := &image{
> -               base:     base,
> -               manifest: m.DeepCopy(),
> +               base:        base,
> +               annotations: annotations,
>         }
>  
>         return image, nil
> ```
> 
> One thing I'm a little worried about is that if you run this command on a manifest list, I think it will turn that manifest list into an image, because we just call `crane.Pull`. We don't want to do that...
> 
> Give me one second, I want to implement something that will let us check if something is an image.

thank you so much for this, I really appreciate it, I updated the PR according to your feedback. 🙋🏻‍♂️😋

**Comment by developer-guy** [comment]
> > Give me one second, I want to implement something that will let us check if something is an image.
> 
> #1057

where should we do this validation, would you like to help me if you don't mind? 🙏

**Comment by jonjohnsonjr** [comment]
> where should we do this validation

So at the beginning of the command, before we `crane.Pull`, we want to check to make sure we're not trying to mutate annotations on an index, because that won't work... Something like:

```go
if len(anntns) != 0 {
  desc, err := crane.Head(ref, *options...)
  if err != nil {
    log.Fatalf("checking %s: %v", ref, err)
  }
  if desc.MediaType.IsIndex() {
    log.Fatalf("mutating annotations on an index is not yet supported")
  }
}
```

**Comment by developer-guy** [comment]
> > where should we do this validation
> 
> So at the beginning of the command, before we `crane.Pull`, we want to check to make sure we're not trying to mutate annotations on an index, because that won't work... Something like:
> 
> ```go
> if len(anntns) != 0 {
>   desc, err := crane.Head(ref, *options...)
>   if err != nil {
>     log.Fatalf("checking %s: %v", ref, err)
>   }
>   if desc.MediaType.IsIndex() {
>     log.Fatalf("mutating annotations on an index is not yet supported")
>   }
> }
> ```

done @jonjohnsonjr in the commit #de9902f.

**Comment by jonjohnsonjr** [comment]
> can you please tell me why we are doing this, and what is the index thing actually, thank you

Sure, let's look at `ubuntu:latest` on DockerHub:

https://explore.ggcr.dev/?image=ubuntu@sha256:aba80b77e27148d99c034a987e7da3a287ed455390352663418c0f2ed40417fe

This is a multi-platform image, AKA an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md), which I often shorten to just "index". Per the spec, an image index can have annotations, so it would be reasonable to expect calling `crane mutate -a "hello=world" ubuntu` to work and add an annotation to the multi-platform image.

However, this will not work! It would instead add the annotation to one of it's children -- the linux/amd64 image:

https://explore.ggcr.dev/?image=index.docker.io/library/ubuntu@sha256:376209074d481dca0a9cf4282710cd30a9e7ff402dea8261acdaaf57a18971dd&mt=application/vnd.docker.distribution.manifest.v2+json

Why? Because `crane.Pull` returns a `v1.Image`, which means it has to pick one of the child images to pull. There's a `--platform` flag and `crane.WithPlatform` option for specifying which platform you should pick.

After annotating the _child_ image, it would update the `ubuntu:latest` tag to point to that _child_ instead of the multi-platform image, which would essentially break any other platforms trying to pull it.

We want to avoid that last part, so if we encounter something that isn't an image, we'll just abort. In the future, we could add something like `mutate.IndexAnnotations` or something to support annotating multi-platform images. This is awful because we bolted on multi-platform stuff after the fact, so it doesn't work that well with the type system. We've had some discussion around how to make this better (https://github.com/google/go-containerregistry/pull/835) but I think we'll need to redesign everything and bump to a v2 to improve things, really :/

**Comment by jonjohnsonjr** [comment]
Looks like the linter is mad, need to run `gofmt`.

**Comment by developer-guy** [comment]
> we'll just abort

> > can you please tell me why we are doing this, and what is the index thing actually, thank you
> 
> Sure, let's look at `ubuntu:latest` on DockerHub:
> 
> https://explore.ggcr.dev/?image=ubuntu@sha256:aba80b77e27148d99c034a987e7da3a287ed455390352663418c0f2ed40417fe
> 
> This is a multi-platform image, AKA an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md), which I often shorten to just "index". Per the spec, an image index can have annotations, so it would be reasonable to expect calling `crane mutate -a "hello=world" ubuntu` to work and add an annotation to the multi-platform image.
> 
> However, this will not work! It would instead add the annotation to one of it's children -- the linux/amd64 image:
> 
> https://explore.ggcr.dev/?image=index.docker.io/library/ubuntu@sha256:376209074d481dca0a9cf4282710cd30a9e7ff402dea8261acdaaf57a18971dd&mt=application/vnd.docker.distribution.manifest.v2+json
> 
> Why? Because `crane.Pull` returns a `v1.Image`, which means it has to pick one of the child images to pull. There's a `--platform` flag and `crane.WithPlatform` option for specifying which platform you should pick.
> 
> After annotating the _child_ image, it would update the `ubuntu:latest` tag to point to that _child_ instead of the multi-platform image, which would essentially break any other platforms trying to pull it.
> 
> We want to avoid that last part, so if we encounter something that isn't an image, we'll just abort. In the future, we could add something like `mutate.IndexAnnotations` or something to support annotating multi-platform images. This is awful because we bolted on multi-platform stuff after the fact, so it doesn't work that well with the type system. We've had some discussion around how to make this better (#835) but I think we'll need to redesign everything and bump to a v2 to improve things, really :/

Omg, thank you so much for this detailed explanation, I got it now 👍

**Comment by developer-guy** [comment]
> Looks like the linter is mad, need to run `gofmt`.

fixed, thank you so much 👍

**Review by jonjohnsonjr — APPROVED** [review]

**Comment by jonjohnsonjr** [comment]
Thanks!
